src/rust/engine/fs/src/store.rs

```diff
@@ -63,76 +66,93 @@ impl Store {
     })
   }
 
-  pub fn store_file_bytes(&self, bytes: Vec<u8>) -> Result<Digest, String> {
+  pub fn store_file_bytes(&self, bytes: Vec<u8>) -> BoxFuture<Digest, String> {
     let len = bytes.len();
-    self.store_bytes(bytes, self.inner.file_store.clone()).map(
-      |fingerprint| Digest(fingerprint, len),
-    )
+    self
+      .store_bytes(bytes, self.inner.file_store.clone())
+      .map(move |fingerprint| Digest(fingerprint, len))
+      .to_boxed()
   }
 
-  fn store_bytes(&self, bytes: Vec<u8>, db: Database) -> Result<Fingerprint, String> {
-    let mut hasher = Sha256::default();
-    hasher.input(&bytes);
-    let fingerprint = Fingerprint::from_bytes_unsafe(hasher.fixed_result().as_slice());
+  fn store_bytes(&self, bytes: Vec<u8>, db: Database) -> CpuFuture<Fingerprint, String> {
```

Comment:
If this were a crate-public API, it would be good to expose `BoxFuture` to avoid leaking implementation details... but since it isn't, avoiding the boxing is just nice.